src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -710,6 +715,12 @@ ExecuteSingleModifyTask(CitusScanState *scanState, Task *task, bool expectResult
 										  taskRequiresTwoPhaseCommit,
 										  startedInTransaction);
 
+	/*
+	 * If we are dealing with partitioned table, we also need to lock its
+	 * partitions.
+	 */
+	LockPartitionRelationsIfPartitioned(relationId, RowExclusiveLock);
```

Comment:
Or maybe we on `CreateDistributedPlan()` we add `relationList` to `MultiPlan` using the `relationRestrictions`. That seems simpler?